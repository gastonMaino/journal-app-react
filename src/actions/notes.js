import swal from 'sweetalert2';


import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
        dispatch(noteAdd(doc.id, newNote));
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const noteAdd = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        }

        const noteFirestore = { ...note };
        delete noteFirestore.id

        db.doc(`${uid}/journal/notes/${note.id}`).update(noteFirestore);

        dispatch(refreshNote(note.id, noteFirestore));

        swal.fire('Saved', note.title, 'success');
    }
}



export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active } = getState().notes

        swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                swal.showLoading();
            }
        })

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        swal.close();

        dispatch(startSaveNote(active))

    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));

        swal.fire('Completed', 'successfully note deleted', 'success')
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})