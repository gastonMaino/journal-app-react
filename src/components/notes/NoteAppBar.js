import React, { useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes'

export const NoteAppBar = () => {
    const dispatch = useDispatch()
    const { active } = useSelector(state => state.notes)

    const inputFile = useRef(null);

    const handleSave = () => {
        dispatch(startSaveNote(active))
    }

    const handlePictureClick = () =>{
        inputFile.current.click();
    }

    const handleFileChange = (e) => {
        const file = e.target.files[0];

        if( file ) {
            dispatch(startUploading(file))
        }
    }


    return (
        <div className='notes__appbar'>
            <time>June 12 2021</time>

            <input 
                type='file'
                style={{display: 'none'}}
                ref={inputFile}
                onChange={handleFileChange}
            />

            <div>
                <button
                    className='btn'
                    onClick={handlePictureClick}
                >
                    picture
                </button>
                <button
                    className='btn'
                    onClick={handleSave}
                >
                    save
                </button>
            </div>
        </div>
    )
}
