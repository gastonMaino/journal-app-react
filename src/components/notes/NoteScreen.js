import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';


import { activeNote, startDeleting } from '../../actions/notes';
import { NoteAppBar } from './NoteAppBar';
import { useForm } from '../../hooks/useForm/useForm';

export const NoteScreen = () => {
    const dispatch = useDispatch();

    const { active } = useSelector(state => state.notes)
    const [values, handleInputChange, reset] = useForm(active);
    const { title, body, id } = values;
    const activeId = useRef(active.id);

    useEffect(() => {
        if (active.id !== activeId.current) {
            reset(active);
            activeId.current = active.id
        }
    }, [active, reset])

    useEffect(() => {
        dispatch(activeNote(values.id, { ...values }));
    }, [values, dispatch])

    const handleDelete = () => {
        dispatch(startDeleting(id));
    }

    return (
        <div className='notes__main-content'>
            <NoteAppBar />

            <div className='notes__content'>
                <input
                    type='text'
                    placeholder='some awesome title'
                    className='notes__title-input'
                    name='title'
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder='what happened today?'
                    className='notes__textarea'
                    name='body'
                    value={body}
                    onChange={handleInputChange}
                ></textarea>
            </div>

            {
                (active.url) &&
                <div className='notes__image'>
                    <img src={active.url} alt='bird cage' />
                </div>
            }

            <button
                className='btn btn-danger'
                onClick={handleDelete}
            >
                delete
            </button>

        </div>
    )
}
