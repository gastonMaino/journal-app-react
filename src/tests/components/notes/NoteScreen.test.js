import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from "enzyme/build"
import { Provider } from 'react-redux';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        name: 'gaston'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 1234,
            title: 'hi',
            body: 'world',
            date: 0
        }
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>
);




describe('test in NoteScreen', () => {
    test('should match to snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should shot activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'update note'
            }
        })

        expect(activeNote).toHaveBeenLastCalledWith(1234, {
            body: 'world',
            title: 'update note',
            id:1234,
            date: 0 
        });
    })
    

})
