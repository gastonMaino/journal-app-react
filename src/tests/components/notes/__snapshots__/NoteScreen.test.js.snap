// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in NoteScreen should match to snapshot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <time>
      June 12 2021
    </time>
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="some awesome title"
      type="text"
      value="hi"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="what happened today?"
      value="world"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    delete
  </button>
</div>
`;
